src/environmentd/tests/sql.rs

```diff
@@ -1961,3 +1959,58 @@ fn test_idle_in_transaction_session_timeout() -> Result<(), Box<dyn Error>> {
 
     Ok(())
 }
+
+#[test]
+fn test_coord_startup_blocking() -> Result<(), Box<dyn Error>> {
+    let initial_time = 0;
+    let now = Arc::new(Mutex::new(initial_time));
+    let now_fn = {
+        let timestamp = Arc::clone(&now);
+        NowFn::from(move || *timestamp.lock().unwrap())
+    };
+    let data_dir = tempfile::tempdir()?;
+    let config = util::Config::default()
+        .with_now(now_fn)
+        .data_directory(data_dir.path());
+
+    // Start 3 servers and reserve the first 3 timestamp ranges.
+    {
+        let server = util::start_server(config.clone())?;
+        let mut client = server.connect(postgres::NoTls)?;
+
+        client.query("SELECT 1", &[])?;
+    };
+    {
+        let server = util::start_server(config.clone())?;
+        let mut client = server.connect(postgres::NoTls)?;
+
+        client.query("SELECT 1", &[])?;
+    };
+    {
+        let server = util::start_server(config.clone())?;
+        let mut client = server.connect(postgres::NoTls)?;
+
+        client.query("SELECT 1", &[])?;
+    };
+
+    let (tx, rx) = std::sync::mpsc::channel();
+    std::thread::spawn(move || {
+        let server = util::start_server(config.clone()).unwrap();
+        let mut client = server.connect(postgres::NoTls).unwrap();
+
+        client.query("SELECT 1", &[]).unwrap();
+        tx.send(()).unwrap();
+    });
+
+    let server_started = Retry::default()
+        .max_duration(Duration::from_secs(1))
+        .retry(|_| rx.try_recv());
+    assert!(server_started.is_err(), "server should be blocked");
```

Comment:
Can you assert that the error message contains some expected string? Want to make sure we are failing for a vaguely correct reason.

Reply:
How do you capture the logs in the test? `server_started` is just going to be a `TryRecvError` (https://doc.rust-lang.org/std/sync/mpsc/struct.Receiver.html#method.try_recv) because the channel is empty.

Reply:
Can you do `tx.send(client.query("select 1", &[]).map(|_| ()))` to send the query error through the channel?

Reply:
It won't be a query error, the task will just be blocked on `let server = util::start_server(config.clone()).unwrap();` as it waits for the server to start up without ever making it to the query.

Reply:
Ok, then don't worry about it.